### ISSUE TITLE
memory reported by lldb does not match memory reported by `top`

### ISSUE THREAD

**Issue by mabead** [dotnet/corefx#28810]
We have memory issue with a .net core 2.0.5 service that runs in a linux container. The service consumes too much memory. After some struggling, I was able to create a memory dump using the following commands:

```
echo "kernel.yama.ptrace_scope=0" | tee -a /etc/sysctl.conf
sysctl -p
gcore -o /tmp/memory_dump 1
```
Where `1` is the PID of the process that consumes too much memory. I also executed the `top` and `htop` commands before generating this dump. Here's what it looked like:

![image](https://user-images.githubusercontent.com/4930754/38312090-d7f8f8b2-37ee-11e8-92c2-934c68a3a9f5.png)

![image](https://user-images.githubusercontent.com/4930754/38312094-db905862-37ee-11e8-8803-aa97e52f8c5c.png)

If I understand properly the output of `top`, the application uses about 1.47 GB of memory. This is also what [prometheus](https://prometheus.io/) reports. 

I then loaded my dump file in LLDB 3.6. Surprisingly, it reports that only 47 **MB** are used. Here's a partial output of `sos EEHeap`:

```
(lldb) sos EEHeap
Loader Heap:
--------------------------------------
System Domain:     00007f33b3160ce0
LowFrequencyHeap:  00007F3338EC0000(3000:3000) 00007F333B410000(10000:10000) 00007F333D5C0000(10000:10000) 00007F333D610000(10000:10000) 00007F333FD40000(10000:10000) 00007F3341F70000(10000:10000) Size: 0x53000 (339968) bytes.
HighFrequencyHeap: 00007F3338EC6000(7000:7000) 00007F333B360000(10000:10000) 00007F333D5A0000(10000:10000) 00007F333D5B0000(10000:10000) 00007F333D5F0000(10000:10000) 00007F333D750000(10000:10000) 00007F333EC90000(10000:10000) 00007F333F1D0000(10000:10000) 00007F333FEA0000(10000:10000) 00007F33413A0000(10000:10000) 00007F3341AA0000(10000:10000) 00007F33421A0000(10000:10000) Size: 0xb7000 (749568) bytes.
StubHeap:          00007F3338ECD000(3000:3000) Size: 0x3000 (12288) bytes.
Virtual Call Stub Heap:
  IndcellHeap:     00007F3338F70000(6000:6000) Size: 0x6000 (24576) bytes.
  LookupHeap:      00007F3338F7C000(4000:4000) Size: 0x4000 (16384) bytes.
  ResolveHeap:     00007F3338FA6000(3a000:3a000) Size: 0x3a000 (237568) bytes.
  DispatchHeap:    00007F3338F80000(26000:26000) Size: 0x26000 (155648) bytes.
  CacheEntryHeap:  00007F3338F76000(6000:6000) Size: 0x6000 (24576) bytes.
Total size:        Size: 0x17d000 (1560576) bytes.
--------------------------------------
...
--------------------------------------
Total LoaderHeap size:   Size: 0x24db000 (38645760) bytes.
=======================================
Number of GC Heaps: 1
generation 0 starts at 0x00007F331615CB28
generation 1 starts at 0x00007F3315FA4488
generation 2 starts at 0x00007F3313FFF000
ephemeral segment allocation context: none
         segment             begin         allocated              size
00007F3313FFE000  00007F3313FFF000  00007F33167BC4F8  0x27bd4f8(41669880)
Large object heap starts at 0x00007F3323FFF000
         segment             begin         allocated              size
00007F3323FFE000  00007F3323FFF000  00007F33247708E0  0x7718e0(7805152)
Total Size:              Size: 0x2f2edd8 (49475032) bytes.
------------------------------
GC Heap Size:            Size: 0x2f2edd8 (49475032) bytes.
```

In a similar way if I do `sos DumpHeap -stat` and I sum the output of the `TotalSize` column, I get a total of 49045050 bytes which is pretty similar to the output of `sos EEHeap`. 

**So here's my question:** how is it possible to have such a big difference in memory consumption with `top` and `lldb`? Could it be possible that lldb only analyzes the managed memory which could point to a possible unmanaged memory leak?

Any other theories are welcomed ;)

**Comment by karelz** [comment]
EEHeap is ExecutionEngine Heap, i.e. managed memory only.
`top` reports `VIRT` also as similar number 49MB. Is it possible that `RES` (1.474g) means reserved memory, but not committed? (or whatever the terms are on Linux when you reserve virtual memory block, but don't ask the OS to block real memory for it just yet)

IMO it would be totally understandable if there was a memory spike in your application, using lots of memory, then it gets released. GC would eventually compat and release most of the memory back to the OS, while keeping reserved virtual address space.

cc @janvorli @Maoni0

**Comment by janvorli** [comment]
> top reports VIRT also as similar number 49MB

@karelz The number reported as VIRT is in kB, so it is 4.9GB :-).

**Comment by karelz** [comment]
Oooops 😊

**Comment by janvorli** [comment]
@mabead the RES being 1.4G is really strange. Given the sizes reported by the EEHeap, it would either have to be a native memory allocated by our runtime or by some native component that your app is using either directly or indirectly. I think I've seen similar problem with openssl holding on a huge amount of memory due to the fact that related data structures in the managed code were being held referenced by some app code.

**Comment by mabead** [comment]
@janvorli The investigations that I did this morning also point toward problems with unmanaged memory. We use CoreCompat.System.Drawing which calls libgdi+ (unmanaged code) under the hood. I was also able to easily memory usage increase and exhaustion by exercising this library more intensively than usual.

Our memory issues started when we migrated to .net core 2.0.5. When doing this migration, we had to upgrade CoreCompat.System.DRawing and also libgdi+. 

I will try to replace CoreCompat.System.Drawing by a fully managed library to analyze images.
